File: ios/LowerBackStretching/Sync/FirebaseSyncBackend.swift
import Foundation
import FirebaseAuth
import FirebaseFirestore

/// Real `SyncBackend` against Firebase: anonymous auth + Firestore writes to
/// the per-user tree the rules in `firebase/firestore.rules` were written for
/// (`/users/{uid}/sessions|routines|programProgress|flexibilityTests`).
///
/// Document ids are deterministic (timestamps / local ids), so retried pushes
/// are idempotent upserts rather than duplicates. Every method is best-effort:
/// failures return false/nil and never throw — SwiftData stays canonical, the
/// cloud copy is a mirror. Only assigned to `SyncController.shared.backend`
/// when `GoogleService-Info.plist` is bundled (see `LowerBackStretchingApp`).
struct FirebaseSyncBackend: SyncBackend {

    func signedInUid() async -> String? {
        Auth.auth().currentUser?.uid
    }

    func signInAnonymously() async -> String? {
        do {
            return try await Auth.auth().signInAnonymously().user.uid
        } catch {
            return nil
        }
    }

    func signOut() async {
        try? Auth.auth().signOut()
    }

    private func set(collection: String, docId: String, data: [String: Any]) async -> Bool {
        var uid = await signedInUid()
        if uid == nil { uid = await signInAnonymously() }
        guard let uid else { return false }
        do {
            try await Firestore.firestore()
                .collection("users").document(uid)
                .collection(collection).document(docId)
                .setData(data, merge: true)
            return true
        } catch {
            return false
        }
    }

    func pushSession(
        programId: String,
        dayNumber: Int,
        durationSeconds: Int,
        completedAtEpochMillis: Int64,
        type: String
    ) async -> Bool {
        await set(
            collection: "sessions",
            docId: String(completedAtEpochMillis),
            data: [
                "programId": programId,
                "dayNumber": dayNumber,
                "durationSeconds": durationSeconds,
                "completedAtEpochMillis": completedAtEpochMillis,
                "type": type,
                "platform": "ios",
            ]
        )
    }

    func pushRoutine(
        localId: String,
        name: String,
        stretchIds: [String],
        displayOrder: Int,
        deletedAtEpochMillis: Int64?
    ) async -> Bool {
        var data: [String: Any] = [
            "name": name,
            "stretchIds": stretchIds,
            "displayOrder": displayOrder,
            "platform": "ios",
        ]
        if let deletedAtEpochMillis { data["deletedAtEpochMillis"] = deletedAtEpochMillis }
        return await set(collection: "routines", docId: localId, data: data)
    }

    func pushProgramProgress(
        programId: String,
        currentDay: Int,
        updatedAtEpochMillis: Int64
    ) async -> Bool {
        await set(
            collection: "programProgress",
            docId: programId,
            data: [
                "currentDay": currentDay,
                "updatedAtEpochMillis": updatedAtEpochMillis,
                "platform": "ios",
            ]
        )
    }

    func pushFlexibilityTest(
        recordedAtEpochMillis: Int64,
        sitAndReachCm: Float?,
        toeTouchCm: Float?,
        shoulderReachCm: Float?
    ) async -> Bool {
        var data: [String: Any] = [
            "recordedAtEpochMillis": recordedAtEpochMillis,
            "platform": "ios",
        ]
        if let sitAndReachCm { data["sitAndReachCm"] = sitAndReachCm }
        if let toeTouchCm { data["toeTouchCm"] = toeTouchCm }
        if let shoulderReachCm { data["shoulderReachCm"] = shoulderReachCm }
        return await set(collection: "flexibilityTests", docId: String(recordedAtEpochMillis), data: data)
    }
}

File: android/app/src/main/java/com/lowerbackstretching/sync/FirebaseSyncBackend.kt
package com.lowerbackstretching.sync

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import kotlinx.coroutines.tasks.await

/**
 * Real [SyncBackend] against Firebase: anonymous auth + Firestore writes to the
 * per-user tree the rules in `firebase/firestore.rules` were written for
 * (`/users/{uid}/sessions|routines|programProgress|flexibilityTests`).
 *
 * Document ids are deterministic (timestamps / local ids), so retried pushes
 * are idempotent upserts rather than duplicates. Every method is best-effort:
 * failures return false/null and never throw into callers — sessions stay
 * canonical in Room, the cloud copy is a mirror.
 */
class FirebaseSyncBackend : SyncBackend {
    private val auth get() = FirebaseAuth.getInstance()
    private val db get() = FirebaseFirestore.getInstance()

    override suspend fun signedInUid(): String? = auth.currentUser?.uid

    override suspend fun signInAnonymously(): String? = try {
        auth.signInAnonymously().await().user?.uid
    } catch (_: Exception) {
        null
    }

    override suspend fun signOut() {
        try {
            auth.signOut()
        } catch (_: Exception) {
            // Ignore — signing out a stale session is best-effort.
        }
    }

    private suspend fun set(collection: String, docId: String, data: Map<String, Any?>): Boolean {
        val uid = signedInUid() ?: signInAnonymously() ?: return false
        return try {
            db.collection("users").document(uid).collection(collection)
                .document(docId)
                .set(data.filterValues { it != null }, SetOptions.merge())
                .await()
            true
        } catch (_: Exception) {
            false
        }
    }

    override suspend fun pushSession(
        programId: String,
        dayNumber: Int,
        durationSeconds: Int,
        completedAtEpochMillis: Long,
        type: String,
    ): Boolean = set(
        collection = "sessions",
        docId = completedAtEpochMillis.toString(),
        data = mapOf(
            "programId" to programId,
            "dayNumber" to dayNumber,
            "durationSeconds" to durationSeconds,
            "completedAtEpochMillis" to completedAtEpochMillis,
            "type" to type,
            "platform" to "android",
        ),
    )

    override suspend fun pushRoutine(
        localId: Long,
        name: String,
        stretchIds: List<String>,
        displayOrder: Int,
        deletedAtEpochMillis: Long?,
    ): Boolean = set(
        collection = "routines",
        docId = localId.toString(),
        data = mapOf(
            "name" to name,
            "stretchIds" to stretchIds,
            "displayOrder" to displayOrder,
            "deletedAtEpochMillis" to deletedAtEpochMillis,
            "platform" to "android",
        ),
    )

    override suspend fun pushProgramProgress(
        programId: String,
        currentDay: Int,
        updatedAtEpochMillis: Long,
    ): Boolean = set(
        collection = "programProgress",
        docId = programId,
        data = mapOf(
            "currentDay" to currentDay,
            "updatedAtEpochMillis" to updatedAtEpochMillis,
            "platform" to "android",
        ),
    )

    override suspend fun pushFlexibilityTest(
        recordedAtEpochMillis: Long,
        sitAndReachCm: Float?,
        toeTouchCm: Float?,
        shoulderReachCm: Float?,
    ): Boolean = set(
        collection = "flexibilityTests",
        docId = recordedAtEpochMillis.toString(),
        data = mapOf(
            "recordedAtEpochMillis" to recordedAtEpochMillis,
            "sitAndReachCm" to sitAndReachCm,
            "toeTouchCm" to toeTouchCm,
            "shoulderReachCm" to shoulderReachCm,
            "platform" to "android",
        ),
    )
}
